### PR TITLE
move to CPU after adding noise

### PR DESCRIPTION
**Pull request by Zilinghan**
Resolve the issue raised by @shourya01  in #150

### CODE REVIEW

File: src/appfl/algorithm/globus_compute_client_step_optimizer.py [hunk 1]
```diff
@@ -67,14 +67,18 @@ def update(self, cli_logger):
...
             scale_value = sensitivity / self.epsilon
             super(GlobusComputeClientStepOptim, self).laplace_mechanism_output_perturb(scale_value)
         
+        ## Move the model parameter to CPU (if not) for communication
+        if (self.cfg.device == "cuda"):            
+            for k in self.primal_state:
+                self.primal_state[k] = self.primal_state[k].cpu()
+        
         if self.send_gradient:
             return compute_gradient(original_model, self.model), cli_logger
```
Comment (yim0331) [inline src/appfl/algorithm/globus_compute_client_step_optimizer.py:83]: Did you check if the return object is located in CPU memory?

File: src/appfl/algorithm/globus_compute_client_optimizer.py [hunk 2]
```diff
@@ -65,14 +65,18 @@ def update(self, cli_logger):
...
             scale_value = sensitivity / self.epsilon
             super(GlobusComputeClientOptim, self).laplace_mechanism_output_perturb(scale_value)
         
+        ## Move the model parameter to CPU (if not) for communication
+        if (self.cfg.device == "cuda"):            
+            for k in self.primal_state:
+                self.primal_state[k] = self.primal_state[k].cpu()
+        
         if self.send_gradient:
             return compute_gradient(original_model, self.model), cli_logger
```
Comment (yim0331) [inline src/appfl/algorithm/globus_compute_client_optimizer.py:81]: Did you check if the return object is located in CPU memory?

**Comment by yim0331** [comment]
@Zilinghan Does Globus communication requires the return value from the client update function to move CPU memory? The [client_optimizer.py code line 72-76](https://github.com/APPFL/APPFL/blob/6237d40b2e00d8fbb6669f191632075e31232202/src/appfl/algorithm/client_optimizer.py#L72C9-L76C66) was added to avoid the memory error while the MPI communication. 
@shourya01 Could you check if the return value from the [compute_gradient](https://github.com/APPFL/APPFL/blob/3304836cf777282e4e28cb9e3ff70e79fe55737f/src/appfl/misc/utils.py#L140) function is located in CPU while using CUDA? I believe it will be better to make the consistency of the memory location of the update function. When the self.send_gradient flag is not set, the return value of the update functions are always in CPU.

**Comment by Zilinghan** [comment]
> @Zilinghan Does Globus communication requires the return value from the client update function to move CPU memory? The [client_optimizer.py code line 72-76](https://github.com/APPFL/APPFL/blob/6237d40b2e00d8fbb6669f191632075e31232202/src/appfl/algorithm/client_optimizer.py#L72C9-L76C66) was added to avoid the memory error while the MPI communication. @shourya01 Could you check if the return value from the [compute_gradient](https://github.com/APPFL/APPFL/blob/3304836cf777282e4e28cb9e3ff70e79fe55737f/src/appfl/misc/utils.py#L140) function is located in CPU while using CUDA? I believe it will be better to make the consistency of the memory location of the return value from the update function. When the self.send_gradient flag is not set, the return value of the update function is always in CPU.

Yes, for Globus Compute, it also requires the model to be moved to CPU

**Pull request merged**